SIMD: Extend the enabled targets for Google Highway quicksort

seiko2plus opened the pull request:
This patch extends CPU targets that used to build Highway's quicksort
  to cover PowerPC/VSX2 and ~IBMZ-14~. Also provides two meson options
  to disable Google-highway and Intel-simd-sort.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

seiko2plus committed:
SIMD: Extend the enabled targets for Google Highway quicksort

  This patch extent CPU targets that used to build Highway's quicksort
  to cover PowerPC/VSX2 and IBMZ-14. Also provides two meson options
  to disable Google-highway and Intel-simd-sort.

seiko2plus commented:
The current version of Google Highway seems to break the build on both PowerPC and IBMZ.

Mousius commented:
> The current version of Google Highway seems to break the build on both PowerPC and IBMZ.

Did you try the latest? I know there's been a lot of changes to 16-bit and @jan-wassenberg was planning another release 😸

Mousius commented:
> The [s390 issue](https://github.com/numpy/numpy/actions/runs/7477479830/job/20350255226?pr=25563) appears to be that we aren't linking with targets.cc which defines hwy::Abort.

Is there a way we can not require `targets.cc` for `VQSortStatic` ?

jan-wassenberg commented on numpy/_core/src/npysort/highway_qsort_16bit.dispatch.cpp:
```diff
@@ -1,5 +1,8 @@
-#include "highway_qsort.hpp"
+#ifdef NPY_HAVE_VSX
+    #define HWY_HAVE_FLOAT16 0
```
It's the other way around: Highway already sets HAVE_FLOAT16 to 0 to inform user code when float16 cannot be used. VQSort counts as 'user code'.
What we could do is define VQSORT_ENABLED to 0 in this file if !HWY_HAVE_FLOAT16.

seiko2plus replied:
Thank you for clarifying it, I made some changes to provide a scalar looping fallback when `HWY_HAVE_FLOAT16` is false, and the build passed after.

jan-wassenberg commented:
> > The [s390 issue](https://github.com/numpy/numpy/actions/runs/7477479830/job/20350255226?pr=25563) appears to be that we aren't linking with targets.cc which defines hwy::Abort.
> 
> Is there a way we can not require `targets.cc` for `VQSortStatic` ?

I think it's only required when calling HWY_ABORT/ASSERT/DASSERT, but VQSort does have a few of those, including in VQSortStatic if !VQSORT_ENABLED but VQSortStatic is still called. We could also consider defining hwy::Abort separately in numpy if you really do not want to link with targets.cc, but that seems brittle, right?

Mousius commented:
> > > The [s390 issue](https://github.com/numpy/numpy/actions/runs/7477479830/job/20350255226?pr=25563) appears to be that we aren't linking with targets.cc which defines hwy::Abort.
> > 
> > 
> > Is there a way we can not require `targets.cc` for `VQSortStatic` ?
> 
> I think it's only required when calling HWY_ABORT/ASSERT/DASSERT, but VQSort does have a few of those, including in VQSortStatic if !VQSORT_ENABLED but VQSortStatic is still called. We could also consider defining hwy::Abort separately in numpy if you really do not want to link with targets.cc, but that seems brittle, right?

Yep, I'd rather not define the Highway function in NumPy, both brittle and weird 😸 I was just curious as up until now we've managed to minimize the size by using `VQSortStatic`, and NumPy isn't going to use much of `targets.cc` other than `Abort` due to having its own dynamic dispatch infrastructure. I wonder if alternatively we could extract `Abort` out of `targets.cc`?

seiko2plus committed:
MAINT: Refactor C++ Quicksort into a Separate Header

  Refactored the C++ implementation of quicksort into a separate header.
  This is to ensure its availability as a fallback when SIMD kernel versions of quicksort are not available.

  The refactored quicksort is now encapsulated within the `np::sort` namespace, and is no longer relying on tags.
  Future patches will transition the rest of the sorting functions to utilize C++ operators instead of relying on tags.

  This patch also introduced a new C++ alias `np::SSize`, mapping to `Py_ssize_t`, and
  implemented a new utility function `np::BitScanReverse` for efficiently performing floor(log2(x)) operations,
  offering improved performance over the existing C function `npy_get_msb()`.

seiko2plus commented:
> Did you try the latest? I know there's been a lot of changes to 16-bit and @jan-wassenberg was planning another release 😸

No, I didn't.

> I think it's only required when calling HWY_ABORT/ASSERT/DASSERT, but VQSort does have a few of those, including in VQSortStatic if !VQSORT_ENABLED but VQSortStatic is still called. We could also consider defining hwy::Abort separately in numpy if you really do not want to link with targets.cc, but that seems brittle, right?

After adding `targets.cc` to source tree the build fails on powerpc due to:
```C
../numpy/_core/src/highway/hwy/targets.cc:41:10: fatal error: asm/hwcap.h: No such file or directory
   41 | #include <asm/hwcap.h>
```
It's a bit strange to require header `<asm/hwcap.h>`, `sys/auxv.h` should be used instead.

Also, I have a crush on s390x CI job:
```Bash
numpy/_core/tests/test_multiarray.py ................................... [  0%]
........................................................................ [  0%]
.............................XX......................Fatal Python error: Aborted

Current thread 0x00000040038f3200 (most recent call first):
  File "/numpy/build-install/usr/lib/python3/dist-packages/numpy/_core/fromnumeric.py", line 1056 in sort
  File "/numpy/build-install/usr/lib/python3/dist-packages/numpy/_core/tests/test_multiarray.py", line 2036 in test_sort
  File "/usr/local/lib/python3.10/dist-packages/_pytest/python.py", line 194 in pytest_pyfunc_call
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 77 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 493 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/python.py", line 1792 in runtest
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 169 in pytest_runtest_call
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 77 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 493 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 262 in <lambda>
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 341 in from_call
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 261 in call_runtest_hook
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 222 in call_and_report
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 133 in runtestprotocol
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 114 in pytest_runtest_protocol
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 77 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 493 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 350 in pytest_runtestloop
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 77 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 493 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 325 in _main
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 271 in wrap_session
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 318 in pytest_cmdline_main
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 77 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 493 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py", line 169 in main
  File "/usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py", line 192 in console_main
  File "/usr/local/lib/python3.10/dist-packages/pytest/__main__.py", line 5 in <module>
  File "/usr/lib/python3.10/runpy.py", line 86 in _run_code
  File "/usr/lib/python3.10/runpy.py", line 196 in _run_module_as_main
```

seiko2plus commented:
I wonder if there's a way to avoid abort and raising a Python runtime error instead. Which Highway operations require that kind of fatal error? Memory allocations?

jan-wassenberg commented:
> It's a bit strange to require header <asm/hwcap.h>, sys/auxv.h should be used instead.

We have hit some systems where one or both are required :/ Highway's CMake includes tooling to check which is available,
you can define TOOLCHAIN_MISS_ASM_HWCAP_H to avoid including that header.

> I wonder if alternatively we could extract Abort out of targets.cc?

I like your idea to pull it out, will do.

> I wonder if there's a way to avoid abort and raising a Python runtime error instead. Which Highway operations require that kind of fatal error? Memory allocations?

Actually allocations do not abort. Mainly it's self-consistency checks or "this is a user/logic error".
Interesting idea with runtime error. How would that be accomplished? We could consider registering an abort handler and hwy::Abort calls that if non-null?

seiko2plus committed:
MAINT: Disable VXE Highway Quicksort build due to runtime crush

seiko2plus commented:
>  you can define TOOLCHAIN_MISS_ASM_HWCAP_H to avoid including that header.

It works, thank you.

> Actually allocations do not abort. Mainly it's self-consistency checks or "this is a user/logic error".
Interesting idea with runtime error. 

It's possible by raising runtime error during the load of Highway something similar to:

https://github.com/numpy/numpy/blob/0ecb9b52234366b8114995455753b43effabf3e4/numpy/_core/src/common/npy_cpu_features.c#L52-L57

> Also, I have a crush on s390x CI job:

I ended up disabling the z14 build, it needs more investigation on the native VM, I am not sure if it's a Qemu bug or Highway one.

Mousius commented:
> I like your idea to pull it out, will do.

Hoping you didn't start work on this already: https://github.com/google/highway/pull/1935 😸 

> Actually allocations do not abort. Mainly it's self-consistency checks or "this is a user/logic error".
> Interesting idea with runtime error. How would that be accomplished? We could consider registering an abort handler and hwy::Abort calls that if non-null?

Maybe this is a use case for having the `hwy::Abort` function interface well defined and stable, and allowing defining one within a project if you want different abort behaviour? I'm not sure it's worth adding all the functionality to dynamically set it given the only use-case is NumPy.

seiko2plus commented:
CI errors are not related, pr is ready 4 review/merge.

jan-wassenberg commented:
> It's possible by raising runtime error during the load of Highway something similar to:

I see, thanks.

> Hoping you didn't start work on this already

No indeed, thanks for doing it :)

> Maybe this is a use case for having the hwy::Abort function interface well defined and stable, and allowing defining one within a project if you want different abort behaviour? I'm not sure it's worth adding all the functionality to dynamically set it given the only use-case is NumPy.

Dynamic setting functionality seems very doable and potentially beneficial.
It could be as simple as:

```
typedef void (*AbortFunc)(const char* file, int line, const char* formatted_err);

AbortFunc& GetAbortFunc() {
	static AbortFunc func;
	return func;
}

// When func is called, and does not intend to terminate the program, it must
// also call the return value of this function.
AbortFunc SetAbortFunc(AbortFunc func) {
	const AbortFunc prev = GetAbortFunc();
	GetAbortFunc() = func;
	return prev;
}
```

Returning the previous function would allow callers to see if someone else has already set one, or also pre-abort 'hooks' enabled by each AbortFunc calling the previous one from their registration.
If GetAbortFunc is null, or the function (chain) still hasn't exited, we resume with the current implementation.
What do you think?

Mousius commented on meson_options.txt:
```diff
@@ -16,6 +16,10 @@ option('mkl-threading', type: 'string', value: 'auto',
+option('disable-highway', type: 'boolean', value: false,
+        description: 'Disables SIMD-optimized operations related to Google Highway(quicksort)')
+option('disable-intel-sort', type: 'boolean', value: false,
+        description: 'Disables SIMD-optimized operations related to Intel-simd-sort(quicksort, argsort, ...)')
```
```suggestion
        description: 'Disables SIMD-optimized operations related to x86-simd-sort(quicksort, argsort, ...)')
```

Mousius commented on meson_options.txt:
```diff
@@ -16,6 +16,10 @@ option('mkl-threading', type: 'string', value: 'auto',
         description: 'MKL threading method, one of: `seq`, `iomp`, `gomp`, `tbb`')
 option('disable-svml', type: 'boolean', value: false,
         description: 'Disable building against SVML')
+option('disable-highway', type: 'boolean', value: false,
+        description: 'Disables SIMD-optimized operations related to Google Highway(quicksort)')
+option('disable-intel-sort', type: 'boolean', value: false,
+        description: 'Disables SIMD-optimized operations related to Intel-simd-sort(quicksort, argsort, ...)')
```
Do we need to list the algorithms here?

seiko2plus replied:
> Do we need to list the algorithms here?

Not necessary but why not give a small brief to the end user on which operations are related to?

Mousius replied:
Because it's so far from the code, I expect we'll forget to update it 😿 I don't mind either way though, it is *sort* of helpful (pun intended)

seiko2plus replied:
Good point, done!

Mousius reviewed (approved):
Minor suggestions but otherwise this looks good to me.

r-devulap reviewed (approved):
Apart from the duplication of quicksort function, LGTM.

seiko2plus commented on numpy/_core/src/npysort/quicksort.hpp:
```diff
@@ -0,0 +1,96 @@
+
+// NUMERIC SORTS
+template <typename T>
+inline void Quick(T *start, SSize num)
```
>  . Could we reuse this implementation in quicksort.cpp?

No, we need to move into a separate header which I did but with some changes as I mentioned within commit log:

> Refactored the C++ implementation of quicksort into a separate header.
  This is to ensure its availability as a fallback when SIMD kernel versions of quicksort are not available.

> The refactored quicksort is now encapsulated within the `np::sort` namespace, and is no longer relying on tags.
  Future patches will transition the rest of the sorting functions to utilize C++ operators instead of relying on tags.

> This patch also introduced a new C++ alias `np::SSize`, mapping to `Py_ssize_t`, and
  implemented a new utility function `np::BitScanReverse` for efficiently performing floor(log2(x)) operations,
  offering improved performance over the existing C function `npy_get_msb()`.

mattip commented:
Is it possible to run benchmarks somewhere to prove this is working? Is there any effect on benchmarks on other architectures (there shouldn't be, just looking for confirmation)?

seiko2plus commented:
> Is it possible to run benchmarks somewhere to prove this is working?

Currently only PowerPC64 is affected by this patch, here a benchmark but on unstable VM:

<details>
<summary>Power9 benchmark</summary>

```Bash
[100.00%] ··· bench_function_base.Sort.time_sort_worst                                                                                                                                                     186±0.6ms
| Change   | Before [7ed86298]    | After [71e9eb30] <highway_extend_sort_targets>   |   Ratio | Benchmark (Parameter)                                                                    |
|----------|----------------------|--------------------------------------------------|---------|------------------------------------------------------------------------------------------|
| +        | 197±0.4μs            | 801±3μs                                          |    4.07 | bench_function_base.Sort.time_sort('quick', 'int64', ('ordered',))                       |
| +        | 229±0.8μs            | 850±6μs                                          |    3.71 | bench_function_base.Sort.time_sort('quick', 'float64', ('ordered',))                     |
| +        | 302±1μs              | 800±2μs                                          |    2.65 | bench_function_base.Sort.time_sort('quick', 'int64', ('reversed',))                      |
| +        | 338±3μs              | 851±2μs                                          |    2.52 | bench_function_base.Sort.time_sort('quick', 'float64', ('reversed',))                    |
| +        | 195±1μs              | 289±3μs                                          |    1.49 | bench_function_base.Sort.time_sort('quick', 'uint32', ('ordered',))                      |
| +        | 197±0.6μs            | 283±2μs                                          |    1.43 | bench_function_base.Sort.time_sort('quick', 'int32', ('ordered',))                       |
| +        | 661±0.7μs            | 881±1μs                                          |    1.33 | bench_function_base.Sort.time_sort('quick', 'int64', ('sorted_block', 1000))             |
| +        | 732±2μs              | 930±4μs                                          |    1.27 | bench_function_base.Sort.time_sort('quick', 'float64', ('sorted_block', 1000))           |
| +        | 244±1μs              | 307±3μs                                          |    1.26 | bench_function_base.Sort.time_sort('quick', 'float32', ('ordered',))                     |
| +        | 501±10μs             | 594±20μs                                         |    1.19 | bench_function_base.Partition.time_partition('float64', ('sorted_block', 1000), 10)      |
| +        | 505±6μs              | 589±8μs                                          |    1.17 | bench_function_base.Partition.time_partition('float64', ('sorted_block', 1000), 100)     |
| +        | 769±4μs              | 889±3μs                                          |    1.16 | bench_function_base.Sort.time_sort('quick', 'int64', ('sorted_block', 100))              |
| +        | 1.02±0.02ms          | 1.16±0.02ms                                      |    1.14 | bench_function_base.Partition.time_partition('float64', ('sorted_block', 1000), 1000)    |
| +        | 618±4μs              | 697±8μs                                          |    1.13 | bench_function_base.Partition.time_argpartition('float64', ('sorted_block', 1000), 100)  |
| +        | 1.13±0.01ms          | 1.28±0.02ms                                      |    1.13 | bench_function_base.Partition.time_argpartition('float64', ('sorted_block', 1000), 1000) |
| +        | 213±2μs              | 240±4μs                                          |    1.13 | bench_function_base.Partition.time_partition('int16', ('sorted_block', 100), 100)        |
| +        | 760±2μs              | 859±2μs                                          |    1.13 | bench_function_base.Sort.time_sort('quick', 'int64', ('sorted_block', 10))               |
| +        | 605±10μs             | 671±4μs                                          |    1.11 | bench_function_base.Partition.time_partition('float64', ('sorted_block', 100), 10)       |
| +        | 600±5μs              | 667±4μs                                          |    1.11 | bench_function_base.Partition.time_partition('float64', ('sorted_block', 100), 100)      |
| +        | 212±0.7μs            | 234±5μs                                          |    1.11 | bench_function_base.Partition.time_partition('int16', ('sorted_block', 100), 10)         |
| +        | 333±0.4μs            | 370±2μs                                          |    1.11 | bench_function_base.Sort.time_argsort('merge', 'int32', ('sorted_block', 10))            |
| +        | 848±1μs              | 940±9μs                                          |    1.11 | bench_function_base.Sort.time_sort('quick', 'float64', ('sorted_block', 100))            |
| +        | 628±6μs              | 690±10μs                                         |    1.1  | bench_function_base.Partition.time_argpartition('float64', ('sorted_block', 1000), 10)   |
| +        | 561±2μs              | 617±7μs                                          |    1.1  | bench_function_base.Partition.time_partition('int16', ('sorted_block', 10), 100)         |
| +        | 615±7μs              | 662±10μs                                         |    1.08 | bench_function_base.Partition.time_partition('float64', ('sorted_block', 100), 1000)     |
| +        | 236±1μs              | 256±0.5μs                                        |    1.08 | bench_function_base.Sort.time_sort('merge', 'float64', ('sorted_block', 100))            |
| +        | 576±10μs             | 616±6μs                                          |    1.07 | bench_function_base.Partition.time_partition('int16', ('sorted_block', 10), 10)          |
| +        | 582±10μs             | 623±3μs                                          |    1.07 | bench_function_base.Partition.time_partition('int16', ('sorted_block', 10), 1000)        |
| +        | 1.03±0ms             | 1.11±0.01ms                                      |    1.07 | bench_function_base.Sort.time_sort('quick', 'float16', ('random',))                      |
| +        | 950±6μs              | 1.00±0.01ms                                      |    1.06 | bench_function_base.Sort.time_sort('quick', 'float16', ('sorted_block', 10))             |
| +        | 78.3±0.2μs           | 82.5±3μs                                         |    1.05 | bench_function_base.Sort.time_sort('heap', 'float32', ('uniform',))                      |
| -        | 919±6μs              | 873±2μs                                          |    0.95 | bench_function_base.Partition.time_partition('int32', ('sorted_block', 10), 10)          |
| -        | 936±5μs              | 886±3μs                                          |    0.95 | bench_function_base.Partition.time_partition('int64', ('sorted_block', 10), 1000)        |
| -        | 500±7μs              | 472±3μs                                          |    0.94 | bench_function_base.Partition.time_partition('float32', ('sorted_block', 1000), 100)     |
| -        | 940±5μs              | 888±1μs                                          |    0.94 | bench_function_base.Partition.time_partition('int32', ('sorted_block', 10), 1000)        |
| -        | 302±1μs              | 284±0.7μs                                        |    0.94 | bench_function_base.Sort.time_sort('quick', 'int32', ('reversed',))                      |
| -        | 761±10μs             | 692±9μs                                          |    0.91 | bench_function_base.Partition.time_argpartition('float32', ('sorted_block', 1000), 100)  |
| -        | 415±8μs              | 377±2μs                                          |    0.91 | bench_function_base.Sort.time_argsort('merge', 'float16', ('sorted_block', 100))         |
| -        | 201±0.4μs            | 182±2μs                                          |    0.91 | bench_function_base.Sort.time_sort('quick', 'int16', ('ordered',))                       |
| -        | 557±5μs              | 493±10μs                                         |    0.89 | bench_function_base.Partition.time_partition('int64', ('sorted_block', 100), 10)         |
| -        | 546±1μs              | 489±6μs                                          |    0.89 | bench_function_base.Partition.time_partition('int64', ('sorted_block', 100), 100)        |
| -        | 558±10μs             | 482±4μs                                          |    0.86 | bench_function_base.Partition.time_partition('int64', ('sorted_block', 100), 1000)       |
| -        | 364±1μs              | 307±2μs                                          |    0.84 | bench_function_base.Sort.time_sort('quick', 'float32', ('reversed',))                    |
| -        | 186±0.6ms            | 112±0.8ms                                        |    0.6  | bench_function_base.Sort.time_sort_worst                                                 |
| -        | 303±1μs              | 180±0.4μs                                        |    0.59 | bench_function_base.Sort.time_sort('quick', 'int16', ('reversed',))                      |
| -        | 645±2μs              | 325±0.9μs                                        |    0.5  | bench_function_base.Sort.time_sort('quick', 'uint32', ('sorted_block', 1000))            |
| -        | 702±2μs              | 324±2μs                                          |    0.46 | bench_function_base.Sort.time_sort('quick', 'int32', ('sorted_block', 1000))             |
| -        | 807±3μs              | 343±1μs                                          |    0.42 | bench_function_base.Sort.time_sort('quick', 'float32', ('sorted_block', 1000))           |
| -        | 754±1μs              | 316±0.7μs                                        |    0.42 | bench_function_base.Sort.time_sort('quick', 'uint32', ('sorted_block', 100))             |
| -        | 296±30μs             | 122±0.9μs                                        |    0.41 | bench_function_base.Sort.time_argsort('merge', 'int16', ('random',))                     |
| -        | 757±2μs              | 308±0.7μs                                        |    0.41 | bench_function_base.Sort.time_sort('quick', 'uint32', ('sorted_block', 10))              |
| -        | 814±1μs              | 308±1μs                                          |    0.38 | bench_function_base.Sort.time_sort('quick', 'int32', ('sorted_block', 10))               |
| -        | 819±3μs              | 315±0.8μs                                        |    0.38 | bench_function_base.Sort.time_sort('quick', 'int32', ('sorted_block', 100))              |
| -        | 838±1μs              | 308±0.7μs                                        |    0.37 | bench_function_base.Sort.time_sort('quick', 'uint32', ('random',))                       |
| -        | 936±3μs              | 337±0.9μs                                        |    0.36 | bench_function_base.Sort.time_sort('quick', 'float32', ('sorted_block', 100))            |
| -        | 952±3μs              | 328±0.5μs                                        |    0.34 | bench_function_base.Sort.time_sort('quick', 'float32', ('sorted_block', 10))             |
| -        | 906±2μs              | 309±1μs                                          |    0.34 | bench_function_base.Sort.time_sort('quick', 'int32', ('random',))                        |
| -        | 1.06±0ms             | 327±1μs                                          |    0.31 | bench_function_base.Sort.time_sort('quick', 'float32', ('random',))                      |
| -        | 738±2μs              | 204±0.8μs                                        |    0.28 | bench_function_base.Sort.time_sort('quick', 'int16', ('sorted_block', 1000))             |
| -        | 848±3μs              | 201±4μs                                          |    0.24 | bench_function_base.Sort.time_sort('quick', 'int16', ('sorted_block', 10))               |
| -        | 203±3μs              | 45.9±0.2μs                                       |    0.23 | bench_function_base.Sort.time_sort('quick', 'float64', ('uniform',))                     |
| -        | 858±1μs              | 194±1μs                                          |    0.23 | bench_function_base.Sort.time_sort('quick', 'int16', ('sorted_block', 100))              |
| -        | 952±5μs              | 196±0.7μs                                        |    0.21 | bench_function_base.Sort.time_sort('quick', 'int16', ('random',))                        |
| -        | 211±0.5μs            | 31.4±0.05μs                                      |    0.15 | bench_function_base.Sort.time_sort('quick', 'float32', ('uniform',))                     |
| -        | 159±2μs              | 20.3±0.2μs                                       |    0.13 | bench_function_base.Sort.time_sort('quick', 'int64', ('uniform',))                       |
| -        | 166±2μs              | 14.3±0.1μs                                       |    0.09 | bench_function_base.Sort.time_sort('quick', 'int32', ('uniform',))                       |
| -        | 157±1μs              | 14.4±0.1μs                                       |    0.09 | bench_function_base.Sort.time_sort('quick', 'uint32', ('uniform',))                      |
| -        | 175±2μs              | 11.2±0.2μs                                       |    0.06 | bench_function_base.Sort.time_sort('quick', 'int16', ('uniform',))                  
```
</details>

>  Is there any effect on benchmarks on other architectures (there shouldn't be, just looking for confirmation)?

This patch only limited to PowerPC64.

mattip commented:
> This patch is only limited to Power architecture .

Yes, that is why I don't expect there to be any impact, but it would be good to confirm just in case there is a bug.

> a benchmark but on an unstable VM

Hmm. Does anyone have access to a Power8 machine and can test this?

seiko2plus commented:
> This patch is only limited to Power architecture .

I was mistaken here; this patch affects both x86 and Arm due to adding support for two Meson options that allow disabling each solution, Intel-sort & highway, during the build time. I ran some tests on both platforms to check the sanity of both build options, and it seems to me, as far as I can tell, that both options are working just fine.

> Hmm. Does anyone have access to a Power8 machine and can test this?

I have access to a native VM of Power9; It's good enough to test the changes before/after, but what I mean by unstable is that the performance stats are slower than on bare metal, and any minor changes should not be taken into consideration.

mattip commented:
> both options are working just fine

Thanks for confirming.

Any idea why a handful of Power9 benchmarks are slower?
